src/structures/idt.rs

```diff
@@ -890,10 +890,204 @@ bitflags! {
     }
 }
 
+#[macro_export]
+/// Set a general handler in an [`InterruptDescriptorTable`].
+/// ```
+/// #![feature(abi_x86_interrupt)]
+/// use x86_64::set_general_handler;
+/// use x86_64::structures::idt::{InterruptDescriptorTable, InterruptStackFrame};
+///
+/// let mut idt = InterruptDescriptorTable::new();
+/// fn my_general_handler(
+///    stack_frame: InterruptStackFrame,
+///    index: u8,
+///    error_code: Option<u64>,
+/// ) {
+///     todo!("handle irq {}", index)
+/// }
+///
+/// // set only one entry
+/// set_general_handler!(&mut idt, my_general_handler, 14);
+///
+/// // set a range of entries
+/// set_general_handler!(&mut idt, my_general_handler, 32..64);
+///
+/// // set all entries
+/// set_general_handler!(&mut idt, my_general_handler);
+/// ```
+macro_rules! set_general_handler {
+    ($idt:expr, $handler:ident) => {
+        $crate::set_general_handler!($idt, $handler, 0..=255);
+    };
+    ($idt:expr, $handler:ident, $idx:literal) => {
+        $crate::set_general_handler!($idt, $handler, $idx..=$idx);
+    };
+    ($idt:expr, $handler:ident, $range:expr) => {{
+        fn set_general_handler(
+            idt: &mut $crate::structures::idt::InterruptDescriptorTable,
+            range: impl core::ops::RangeBounds<usize>,
+        ) {
+            $crate::set_general_handler_recursive_bits!(idt, $handler, range);
+        }
+        set_general_handler($idt, $range);
+    }};
+}
+
+#[macro_export]
+#[doc(hidden)]
+/// We can't loop in macros, but we can use recursion.
+/// This macro recursivly adds one more bit to it's arguments until we have 8 bits so that we can call set_general_handler_entry.
+macro_rules! set_general_handler_recursive_bits {
```

Comment:
I'm wondering if we can do this without the recursive bits. Maybe something like:
```rust
#[doc(hidden)]
macro_rules! set_general_handler_range {
    ($idt:expr, $handler:ident, $range:expr) => {{
        // Exceptions with errors
        set_general_handler_entry!(
            $idt, $handler, $range, true,
            8, 10, 11, 12, 13, 14, 17, 30
        );
        // Exceptions without errors
        set_general_handler_entry!(
            $idt, $handler, $range, false,
            0, 1, 2, 3, 4, 5, 6, 7, 9, 16, 18, 19, 20
        );
        // Interrupts (no error code)
        set_general_handler_intr!(
            $idt, $handler, $range, false,
            32, 48, 64, 80, 96, 112, 128, 144, 160, 176, 192, 208, 224, 240
        );
    }}
}

#[doc(hidden)]
macro_rules! set_general_handler_intr {
    ($idt:expr, $handler:ident, $range:expr, $err:tt, $i:tt) => {{
        set_general_handler_entry!(
            $idt, $handler, $range, $err,
            $i+0, $i+1, $i+2, $i+3, $i+4, $i+5, $i+6, $i+7,
            $i+8, $i+9, $i+10, $i+11, $i+12, $i+13, $i+14, $i+15
        );;
    }}
    ($idt:expr, $handler:ident, $range:expr, $err:tt, $base:tt $(, $bases:tt)+) => {{
        // Expand list
        $crate::set_general_handler_intr!($idt, $handler, $range, $err, $base);
        $crate::set_general_handler_intr!($idt, $handler, $range, $err, $(, $bases:tt)+);
    }}
}

#[doc(hidden)]
macro_rules! set_general_handler_entry {
    ($idt:expr, $handler:ident, true, 8) => {{
        // Special handling for double fault handler
    }}
    ($idt:expr, $handler:ident, false, 18) => {{
        // Special handling for machine fault handler
    }}
    ($idt:expr, $handler:ident, true, $idx:expr) => {{
        // Generic implementation for interrupt with an error code
    }}
    ($idt:expr, $handler:ident, false, $idx:expr) => {{
        // Generic implementation for interrupt without an error code
    }}
    // 5 arguments
    ($idt:expr, $handler:ident, $range:expr, $err:tt, $idx:tt) => {{
        // Range check
        if $range.contains(&$idx) {
            $crate::set_general_handler_entry!($idt, $handler, $err, $idx);
        }
    }}
    // 6+ arguments
    ($idt:expr, $handler:ident, $range:expr, $err:tt, $idx:tt $(, $idxs:tt)+) => {{
        // Expand list
        $crate::set_general_handler_entry!($idt, $handler, $range, $err, $idx);
        $crate::set_general_handler_entry!($idt, $handler, $range, $err, $(, $idxs:tt)+);
    }}
}
```

Reply:
What would your implementation of
```rust
// Generic implementation for interrupt with an error code
```
look like? AFAICT there is currently no way the set a handler with an error code without naming it.

Reply:
What specifically don't you like about the recursive bits? What's the advantage of generating the indices using lists of numbers?

Reply:
While the current implementation is a bit intimidating on first sight, it isn't too complex when taking a closer look. So I think we can merge this as is. We can still simplify this in follow-up PRs if we want.